src/catalog/error.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt;
+
+use backtrace::Backtrace;
+
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+    backtrace: Backtrace,
+}
+
+#[derive(Debug)]
+pub(crate) enum ErrorKind {
+    Corruption { detail: String },
+    IdExhaustion,
+    UnknownDatabase(String),
+    UnknownSchema(String),
+    UnknownItem(String),
+    DatabaseAlreadyExists(String),
+    SchemaAlreadyExists(String),
+    ItemAlreadyExists(String),
+    UnacceptableSchemaName(String),
+    ReadOnlySystemSchema(String),
+    UnsatisfiableLoggingDependency { depender_name: String },
+    Storage(rusqlite::Error),
```

Comment:
I think that if this took a `context: Option<String>` that we could use for logging it would be strictly nicer than the other pr.